### ISSUE TITLE
[v2] Mark packages as "external" for static build to make that build faster

### ISSUE THREAD

**Issue by KyleAMathews** [gatsbyjs/gatsby#5605]
ref https://github.com/mapbox/batfish/issues/261

**Comment by pieh** [comment]
I was reading about this earlier today and found https://www.npmjs.com/package/webpack-node-externals

**Comment by KyleAMathews** [comment]
Oh that looks perfect! Nice!

**Comment by pieh** [comment]
I just did some research and here's what I needed to make it work using package I linked before:
https://github.com/pieh/gatsby/commit/0821cc5ac80254449a2fd0ddaa21221d53427e08

This decreased time building bundle for build-html from ~5.5s to ~1.5s for gatsbyjs.org , but it doesn't handle scenario when npm package have `main` field pointing to file that is not handled by node, so potentially we would either need to fork `webpack-node-externals` and add support or just copy part of code from that package to gatsby.

I'm not sure if it's worth investing more time into doing that before v2 release?

Not using `webpack-node-externals` and specyfing few key gatsby dependencies as externals did drop time from ~5.5s to ~4.5s but would need to experiment with it more and rerun tests on this one (this was just quick test to see if I can quickly get similiar results)

**Comment by KyleAMathews** [comment]
Wow, that's a big change. Getting the externals package working isn't a blocker but adding a few big dependencies seems like a nice win. Exciting!

**Comment by KyleAMathews** [comment]
On the main thing, are you talking about packages that'd have node and browser specific entries?

**Comment by pieh** [comment]
> On the main thing, are you talking about packages that'd have node and browser specific entries?

No, just something like that https://github.com/KyleAMathews/typefaces/blob/0bdb2788fb790cec7560de1d4897a0658fca8ac1/packages/spectral/package.json#L5

And that's imported like that on gatsbyjs.org:
https://github.com/gatsbyjs/gatsby/blob/001158ebe5725eed045fd8439c05e8abbc5916e7/www/src/components/layout.js#L23

`webpack-node-externals` doesn't check entry at all - it just checks import path/request string:
https://github.com/liady/webpack-node-externals/blob/cfa1c5b33752fb8cb72360167ae89b23816cdefe/index.js#L39-L50

That's why I hardcode whitelisted `typeface-spectral` in my testing https://github.com/pieh/gatsby/commit/0821cc5ac80254449a2fd0ddaa21221d53427e08#diff-c4777e791f64fbb6c82255341b381dabR462

But now I'm thinking that checking `main` or other entires might not be enough - in case of some UI toolkits users might import components (which would get marked as externals) that later on will import styles which can't be handled by node. So probably safe bet would be to skip  `webpack-node-externals`, manually specify which packages are externals and maybe allow users and plugins to add to that list in `onCreateWebpackConfig`

**Comment by alpgumus** [comment]
@pieh  The external webpack change broke things when used with redux, which requires lodash-es but gatsby latest webpack config has that in externals /^lodash/

**Comment by KyleAMathews** [comment]
How do things break exactly?

**Comment by alpgumus** [comment]
It is failing to build html (develop or build html)

It is complaining about lodash-es package (es6 syntax) as webpack is excluding it.

```
info bootstrap finished - 3.468 s

error Unexpected identifier


  Error: node_modules\lodash-es\isPlainObject.js:1
  (function (exports, require, module, __filename, __dirname) {
 import baseGetTag from './_baseGetTag.js';
               ^^^^^^^^^^
  SyntaxError: Unexpected identifier

  - loader.js:657 Module._compile
    internal/modules/cjs/loader.js:657:28

  - loader.js:700 Object.Module._extensions..js
    internal/modules/cjs/loader.js:700:10
```

**Comment by m-allanson** [comment]
I just ran in to this, this should fix it: https://github.com/gatsbyjs/gatsby/pull/7325

**Comment by m-allanson** [comment]
The fix was published in `2.0.0-beta.102`
